[Merged by Bors] - feat(LinearAlgebra/LinearDisjoint): definition and properties of linearly disjoint of submodules

acmepjz opened the pull request:
This is part 1 of #9651.

We adapt the definitions in <https://en.wikipedia.org/wiki/Linearly_disjoint>.
Let `R` be a commutative ring, `S` be an `R`-algebra (not necessarily commutative).
Two `R`-submodules `M` and `N` in `S` are linearly disjoint,
if the natural `R`-linear map `M ⊗[R] N →ₗ[R] S` (`Submodule.mulMap M N`)
induced by the multiplication in `S` is injective.

The following is the first equivalent characterization of linearly disjointness:

- `Submodule.LinearDisjoint.linearIndependent_left_of_flat`:
  if `M` and `N` are linearly disjoint, if `N` is a flat `R`-module, then for any family of
  `R`-linearly independent elements `{ m_i }` of `M`, they are also `N`-linearly independent,
  in the sense that the `R`-linear map from `ι →₀ N` to `S` which maps `{ n_i }`
  to the sum of `m_i * n_i` (`Submodule.mulLeftMap N m`) is injective.

- `Submodule.LinearDisjoint.of_basis_left`:
  conversely, if `{ m_i }` is an `R`-basis of `M`, which is also `N`-linearly independent,
  then `M` and `N` are linearly disjoint.

Dually, we have:

- `Submodule.LinearDisjoint.linearIndependent_right_of_flat`:
  if `M` and `N` are linearly disjoint, if `M` is a flat `R`-module, then for any family of
  `R`-linearly independent elements `{ n_i }` of `N`, they are also `M`-linearly independent,
  in the sense that the `R`-linear map from `ι →₀ M` to `S` which maps `{ m_i }`
  to the sum of `m_i * n_i` (`Submodule.mulRightMap M n`) is injective.

- `Submodule.LinearDisjoint.of_basis_right`:
  conversely, if `{ n_i }` is an `R`-basis of `N`, which is also `M`-linearly independent,
  then `M` and `N` are linearly disjoint.

The following is the second equivalent characterization of linearly disjointness:

- `Submodule.LinearDisjoint.linearIndependent_mul_of_flat`:
  if `M` and `N` are linearly disjoint, if one of `M` and `N` is flat, then for any family of
  `R`-linearly independent elements `{ m_i }` of `M`, and any family of
  `R`-linearly independent elements `{ n_j }` of `N`, the family `{ m_i * n_j }` in `S` is
  also `R`-linearly independent.

- `Submodule.LinearDisjoint.of_basis_mul`:
  conversely, if `{ m_i }` is an `R`-basis of `M`, if `{ n_i }` is an `R`-basis of `N`,
  such that the family `{ m_i * n_j }` in `S` is `R`-linearly independent,
  then `M` and `N` are linearly disjoint.

Other results:

- `Submodule.LinearDisjoint.symm_of_commute`, `Submodule.linearDisjoint_symm_of_commute`:
  linearly disjoint is symmetric under some commutative conditions.

- `Submodule.linearDisjoint_op`:
  linearly disjoint is preserved by taking multiplicative opposite.

- `Submodule.LinearDisjoint.of_le_left_of_flat`, `Submodule.LinearDisjoint.of_le_right_of_flat`,
  `Submodule.LinearDisjoint.of_le_of_flat_left`, `Submodule.LinearDisjoint.of_le_of_flat_right`:
  linearly disjoint is preserved by taking submodules under some flatness conditions.

- `Submodule.LinearDisjoint.of_linearDisjoint_fg_left`,
  `Submodule.LinearDisjoint.of_linearDisjoint_fg_right`,
  `Submodule.LinearDisjoint.of_linearDisjoint_fg`:
  conversely, if any finitely generated submodules of `M` and `N` are linearly disjoint,
  then `M` and `N` themselves are linearly disjoint.

- `Submodule.LinearDisjoint.bot_left`, `Submodule.LinearDisjoint.bot_right`:
  the zero module is linearly disjoint with any other submodules.

- `Submodule.LinearDisjoint.one_left`, `Submodule.LinearDisjoint.one_right`:
  the image of `R` in `S` is linearly disjoint with any other submodules.

- `Submodule.LinearDisjoint.of_left_le_one_of_flat`,
  `Submodule.LinearDisjoint.of_right_le_one_of_flat`:
  if a submodule is contained in the image of `R` in `S`, then it is linearly disjoint with
  any other submodules, under some flatness conditions.

- `Submodule.LinearDisjoint.not_linearIndependent_pair_of_commute_of_flat`,
  `Submodule.LinearDisjoint.rank_inf_le_one_of_commute_of_flat`:
  if `M` and `N` are linearly disjoint, if one of `M` and `N` is flat, then any two commutative
  elements contained in the intersection of `M` and `N` are not `R`-linearly independent (namely,
  their span is not `R ^ 2`). In particular, if any two elements in the intersection of `M` and `N`
  are commutative, then the rank of the intersection of `M` and `N` is at most one.

- `Submodule.LinearDisjoint.rank_le_one_of_commute_of_flat_of_self`:
  if `M` and itself are linearly disjoint, if `M` is flat, if any two elements in `M`
  are commutative, then the rank of `M` is at most one.

The results with name containing "of_commute" also have corresponding specified versions
assuming `S` is commutative.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

discussion: <https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Linearly.20disjoint>

- [x] depends on: #12498

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,729 @@
+/-- If `M` and `N` are submodules in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] N →ₗ[R] M * N` induced by multiplication in `S`,
+which is surjective (`Submodule.mulMap'_surjective`). -/
+def mulMap' := (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
```
```suggestion
def mulMap' : M ⊗[R] N →ₗ[R] ↥(M * N) :=
  (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
```
Does this work?

erdOne replied:
In general please add a type ascription if possible and add a note explaining the reason otherwise.

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,729 @@
+def mulMap' := (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
+
+@[simp]
+theorem coe_mulMap'_tmul (m : M) (n : N) : (mulMap' M N (m ⊗ₜ[R] n) : S) = m.1 * n.1 := rfl
```
```suggestion
theorem val_mulMap'_tmul (m : M) (n : N) : (mulMap' M N (m ⊗ₜ[R] n) : S) = m.1 * n.1 := rfl
```

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,729 @@
+    LinearMap.surjective_rangeRestrict]
+
+/-- If `N` is a submodule in an algebra `S` over `R`, there is the natural map
+`i(R) ⊗[R] N →ₗ[R] N` induced by multiplication in `S`, here `i : R → S` is the structure map. -/
```
Please mention `lTensorOne` in the docstring.
(Something like "This is promoted to an equiv as `lTensorOne`. Use that instead.")

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,729 @@
+    N.lTensorOne.symm n = ⟨1, one_le.1 (le_refl _)⟩ ⊗ₜ[R] n := rfl
+
+/-- If `M` is a submodule in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] i(R) →ₗ[R] M` induced by multiplication in `S`, here `i : R → S` is the structure map. -/
```
Same here.

erdOne reviewed (commented):
I've reviewed up to `end mulMap` (L297). I think everything until that point can go into a new file `Mathlib.LinearAlgebra.TensorProduct.Submodule`. And it will probably be merged sooner if it is split into a separate PR.

acmepjz commented:
> I've reviewed up to `end mulMap` (L297). I think everything until that point can go into a new file `Mathlib.LinearAlgebra.TensorProduct.Submodule`. And it will probably be merged sooner if it is split into a separate PR.

Added as #12498. Maybe some proof in this PR can be golfed since I added `mulLeftMap_eq_mulRightMap_of_commute` etc. in that PR.

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,550 @@
+
+variable (M N)
+
+theorem of_map_linearIndependent_left' {ι : Type*} (m : Basis ι R M)
```
I think the name should mention `Basis` instead of `linearIndependent`.

acmepjz replied:
Done.

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,550 @@
+  exact M.injective_subtype.comp M.rTensorOne.injective
+
+theorem of_linearDisjoint_finite_left
+    (H : ∀ M' : Submodule R S, M' ≤ M → [Module.Finite R M'] → M'.LinearDisjoint N) :
```
I think `M'.FG` makes more sense here.

acmepjz replied:
Done.

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,550 @@
+  exact of_map_linearIndependent_right' M N n (LinearMap.ker_eq_bot.1 H)
+
+variable {M N} in
+theorem map_linearIndependent_mul_of_flat_left (H : M.LinearDisjoint N) [Module.Flat R M]
```
What is the `map` referring to here?

acmepjz replied:
It means that it maps two linearly independent families to a new linear independent family, which is the product of these two families. It's written in the module docstring in the beginning of the file:

>  if `M` and `N` are linearly disjoint, if one of `M` and `N` is flat, then for any family of
>  `R`-linearly independent elements `{ m_i }` of `M`, and any family of
>  `R`-linearly independent elements `{ n_j }` of `N`, the family `{ m_i * n_j }` in `S` is
>  also `R`-linearly independent.

I guess I should also copy the docstring to each function.

acmepjz replied:
Docstring added.

acmepjz replied:
I see. Do you think it's a good idea to remove `map_` from the name of other results?

erdOne commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,550 @@
+    (a b : ↥(M ⊓ N)) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat_right a b (mul_comm _ _)
+
+theorem not_linearIndependent_pair_of_flat (hf : Module.Flat R M ∨ Module.Flat R N)
+    (a b : ↥(M ⊓ N)) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat hf a b (mul_comm _ _)
+
+theorem not_linearIndependent_pair_of_flat' (hf : Module.Flat R M ∨ Module.Flat R N)
+    {a b : S} (ha : a ∈ M ⊓ N) (hb : b ∈ M ⊓ N) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat' hf ha hb (mul_comm _ _)
```
Do we need to specialize these? I thought they were helper lemmas.

acmepjz replied:
Do you mean `flat` version vs `flat'` version? The `flat` is bundled version (i.e. `a b : ↥(M ⊓ N)`) while the `flat'` version is not (i.e. `a b : S` with the condition that they are `∈ M ⊓ N`). But maybe you are right, I checked the downstream code and it looks like that the `flat'` version is unused. Do you think it's a good idea to remove `flat'` version?

acmepjz replied:
`flat'` version removed.

erdOne reviewed (approved):
LGTM. Some small nitpicks on naming.

acmepjz commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,619 @@
+
+/-- The `Submodule.LinearDisjoint.not_linearIndependent_pair_of_commute_of_flat`
+for commutative rings. -/
+theorem not_linearIndependent_pair_of_flat (hf : Module.Flat R M ∨ Module.Flat R N)
```
It is not essential; there are some corollaries of it, stated using `Module.Flat R M ∨ Module.Flat R N`; it's always possible to rewrite the proof for splitting cases left and right.

But personally I think it's convenient if we have a `or`ed version, just like [IntermediateField.sup_toSubalgebra_of_isAlgebraic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.sup_toSubalgebra_of_isAlgebraic) (sorry that is also contributed by me).

acmepjz commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,619 @@
+  exact ⟨H⟩
+
+/-- The zero module is linearly disjoint with any other submodules. -/
+theorem of_bot_left : (⊥ : Submodule R S).LinearDisjoint N :=
```
I think you're right; will change it.

erdOne commented:
Thanks!
maintainer merge
btw a diff of +600 is quite big for a PR and it makes this PR harder to review. Next time please try to split it up. (IMO +250 is max for a PR, and maybe +400 if it only touches one file, and this is probably on the high end among the consensus of all reviewers)

erdOne commented:
maintainer merge

riccardobrasca commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,619 @@
+
+-/
+
+open scoped Classical TensorProduct
```
```suggestion
open scoped TensorProduct
```
A couple of proof need to be fixed.

riccardobrasca commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,619 @@
+/-- Two submodules `M` and `N` in an algebra `S` over `R` are linearly disjoint if the natural map
+`M ⊗[R] N →ₗ[R] S` induced by multiplication in `S` is injective. -/
+@[mk_iff]
+protected structure LinearDisjoint : Prop where
```
Why `protected`?

acmepjz replied:
Because later we will have `Subalgebra.LinearDisjoint`, `IntermediateField.LinearDisjoint` and `Field.LinearDisjoint`.

riccardobrasca commented on Mathlib/LinearAlgebra/LinearDisjoint.lean:
```diff
@@ -0,0 +1,619 @@
+
+/-- If `M` and `N` are linearly disjoint submodules, then there is the natural isomorphism
+`M ⊗[R] N ≃ₗ[R] M * N` induced by multiplication in `S`. -/
+protected def LinearDisjoint.mulMap (H : M.LinearDisjoint N) : M ⊗[R] N ≃ₗ[R] ↥(M * N) :=
```
```suggestion
protected def LinearDisjoint.mulMap (H : M.LinearDisjoint N) : M ⊗[R] N ≃ₗ[R] (M * N) :=
```

riccardobrasca reviewed (approved):
Thanks a lot for this contribution! I see a lot of `simp_rw`, is there a particular reason?

bors d+

acmepjz commented:
> I see a lot of `simp_rw`, is there a particular reason?

I forgot. Probably because `rw` and `simp only` don't work for items inside something.

riccardobrasca commented:
> > I see a lot of `simp_rw`, is there a particular reason?
> 
> I forgot. Probably because `rw` and `simp only` don't work for items inside something.

OK, no problem.

acmepjz commented:
bors r+
